zipkin-server/src/main/scala/org/http4s/zipkin/server/ZipkinServer.scala

```diff
@@ -0,0 +1,83 @@
+package org.http4s.zipkin.server
+
+import java.time.Instant
+
+import org.http4s.headers.{`X-B3-ParentSpanId`, `X-B3-SpanId`, `X-B3-TraceId`}
+import org.http4s.zipkin.core.algebras.{Clock, Collector, Randomness}
+import org.http4s.zipkin.core.{AnnotationType, Endpoint, ServerIds}
+import org.http4s.zipkin.middleware._
+import org.http4s.{Service, _}
+import org.http4s.zipkin.middleware.server.ZipkinService
+import org.http4s.zipkin.core._
+
+import scalaz.{Kleisli, Scalaz}
+
+object ZipkinServer {
+
+  def apply(
+    collector: Collector,
+    randomness: Randomness,
+    clock: Clock,
+    endpoint: Endpoint
+  )(zipkinService: ZipkinService): HttpService = {
+
+    Service.lift { request =>
+      val name = nameFromRequest(request)
+      for {
+        freshSpanId <- Randomness.getRandomLong(randomness)
+        serverIds = serverIdsFromHeaders(request.headers)(freshSpanId)
+
+        srInfo <- Clock.getInstant(clock).map(serverReceiveInfo(true, name, endpoint, serverIds))
+        _ <- Collector.send(srInfo)(collector)
+
+        responseForOurClient <- zipkinService.run(
+          ServerRequirements(serverIds))(request)
+
+        ssInfo <- Clock.getInstant(clock).map(serverSendInfo(true, name, endpoint, serverIds))
+        _ <- Collector.send(ssInfo)(collector)
+
+      } yield responseForOurClient
+    }
+  }
+
+  // When we are a server on the edge of the cluster,
+  // we generate our own span id and trace id
+  // (which we set to the span id out of convention).
+  def serverIdsFromHeaders(headers: Headers)(freshSpanId: Long): ServerIds = {
+    val generatedIds = ServerIds(
+      spanId = `X-B3-SpanId`(freshSpanId),
+      traceId = `X-B3-TraceId`(freshSpanId),
```

Comment:
separate issue, but note that we're widening X-B3-TraceId to 128 bits. You can set all high bits to zero for interop. This is happening over the rest of the year. https://github.com/openzipkin/b3-propagation/issues/6